src/main/java/io/netty/buffer/api/Send.java

```diff
@@ -56,7 +56,7 @@ public T receive() {
             }
 
             @Override
-            public boolean isInstanceOf(Class<?> cls) {
+            public boolean referentIsInstanceOf(Class<?> cls) {
                 return cls.isAssignableFrom(concreteObjectType);
```

Comment:
This looks reversed. `concreteObjectType.isAssignableFrom(cls)`. I think the ordering is opposite `isInstance()`. (`a.isInstance(b) == true` says b can be assigned to a variable of type a; `a.isAssignableFrom(b)` says a can be assigned to a variable of type b). I noticed because `Send.isSendOf(Object.class, x)` should have always returned true if we are using `instanceof` semantics.

That said, as long as we are using instanceof semantics, the caller should be casting to `Send<? extends X>` instead of `Send<X>`. Although that is more verbose, I'm not sure it actually hurts anything.

Note that changing to class equality `cls.equals(concreteObjectType)` (such that we can cast to `Send<X>`) would be a _very good_ reason not to have a separate composite class type (https://github.com/netty/netty-incubator-buffer-api/issues/47).

Reply:
The same code is in `TransferSend`, and flipping the order there makes the `BufferSendTest` fail.

According to the `isAssignableFrom` javadoc:

> Determines if the class or interface represented by this Class object is either the same as, or is a superclass or superinterface of, the class or interface represented by the specified Class parameter.

So the ordering is correct as is.

If we used class equality, then we wouldn't be able to ask about interfaces of the object, and `Buffer` is an interface, so I don't think that would fly.

I'll share my thoughts on the composite class type on that issue.

Reply:
I missed that `assertFalse(Send.isSendOf(Object.class, new Object()));` was passing `new Object()` and not bufferSend/bufferRefSend.

I had thought that isAssignableFrom and isInstance behaved in the same order, and reads the docs once and thought it agreed with me. But then I looked _a second time_ and mis-read the second paragraph of the text and re-read it all and then convinced myself of the other direction. /sigh

> If we used class equality, then we wouldn't be able to ask about interfaces of the object, and Buffer is an interface, so I don't think that would fly.

No, you can still do it for Buffer. But the object would need to know the type of generic it has. That would mean it couldn't be a default method on Rc, but otherwise would be fine.

As we write more code, we may want to keep watch of whether we want a method that will do the cast for you after checking the provided Class. That could return a proper-generics `Send<? extends X>` as well as silence unchecked cast warnings.